### PR TITLE
Move app launcher stop before unregister app

### PR DESCRIPTION
**Pull request by ychernysheva**
Fixes [FORDTCN-1722](https://luxproject.luxoft.com/jira/browse/FORDTCN-1722)

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Summary
The applauncher has to be stopped before appropriate application will be unregistered. Otherwise it could lead to core crash, when Launch controller will try to run already non existed application

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/src/app_launch/app_launch_ctrl_impl.cc [hunk 1]
```diff
@@ -83,6 +83,8 @@ bool HmiLevelSorter(const std::pair<int32_t, ApplicationDataPtr>& lval,
 
 void AppLaunchCtrlImpl::OnDeviceConnected(const std::string& device_mac) {
   SDL_LOG_AUTO_TRACE();
+  sync_primitives::AutoLock lock(launch_ctrl_lock_);
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/src/app_launch/app_launch_ctrl_impl.cc:86]: @ychernysheva what is the purpose of that lock?

File: src/components/application_manager/src/app_launch/device_apps_launcher.cc [hunk 2]
```diff
@@ -157,6 +158,23 @@ bool DeviceAppsLauncherImpl::StopLaunchingAppsOnDevice(
...
+  free_launchers_.push_back(launcher);
+}
+
+void DeviceAppsLauncherImpl::StopLaunchingAppsOnAllDevices() {
+  SDL_LOG_AUTO_TRACE();
+  sync_primitives::AutoLock lock(launchers_lock_);
+
+  std::for_each(works_launchers_.begin(),
+                works_launchers_.end(),
+                boost::bind(&DeviceAppsLauncherImpl::StopLauncher, this, _1));
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/src/app_launch/device_apps_launcher.cc:173]: @ychernysheva I think you can use bind from `std` library now

Comment (ychernysheva) [reply src/components/application_manager/src/app_launch/device_apps_launcher.cc]: Replaced with lambda in 792b6e0

**Review by AKalinich-Luxoft — APPROVED** [review]

**Review by RomanReznichenkoLuxoft — APPROVED** [review]

File: src/components/application_manager/src/app_launch/app_launch_ctrl_impl.cc (hunk 1 above)
Comment (ychernysheva) [reply src/components/application_manager/src/app_launch/app_launch_ctrl_impl.cc:86]: @AKalinich-Luxoft It seems that the main purpose of this lock is avoiding races, related to free_launchers_ container.

**Review by VadymLuchko — APPROVED** [review]

**Pull request closed without merging**